### PR TITLE
fix(stats): improve stats app resiliency

### PR DESCRIPTION
**Pull request by forbesus**
ref #330

### CODE REVIEW

**Comment by forbesus** [comment]
Hello @baktun14 would you check it?

File: apps/api/src/services/db/statsService.ts
Comment (Redm4x) [inline apps/api/src/services/db/statsService.ts]: I think there is no need to try/catch in here. We do not want to show 0 everywhere as it would cause confusion. If an exception is thrown in `getDashboardData` we should just let it reach the catch block in `dashboardData.ts` and it will be handled there correctly.

File: apps/api/src/services/external/apiNodeService.ts
Comment (Redm4x) [inline apps/api/src/services/external/apiNodeService.ts]: In the frontend (`Dashboard.tsx`) the tiles are only hidden if those values are `undefined`. Since they default to `null` here that means if some of them fail to load the tiles are still shown with values of 0.

File: apps/api/src/routes/v1/dashboardData.ts
Comment (Redm4x) [inline apps/api/src/routes/v1/dashboardData.ts]: Please use the new logger instead of `console.error`. You can instantiate it at the top of the file like so:
`const logger = new LoggerService({ context: "DashboardDataEndpoint" });`

**Review by Redm4x — CHANGES_REQUESTED** [review]

File: apps/api/src/routes/v1/dashboardData.ts
Comment (Redm4x) [reply apps/api/src/routes/v1/dashboardData.ts]: It would also be a good idea to send those exceptions to Sentry.
```
import * as Sentry from "@sentry/node";
```
```
Sentry.captureException(error);
```

File: apps/api/src/services/db/statsService.ts
Comment (ygrishajev) [inline apps/api/src/services/db/statsService.ts]: I think props `prev`/`next`, `prev`/`current` would better describe the contents. `compare` makes me wonder what that might be, it's also a verb which complicates it more. `now` sounds like a timestamp value, sounds confusing.

File: apps/api/src/routes/v1/dashboardData.ts
Comment (ygrishajev) [inline apps/api/src/routes/v1/dashboardData.ts]: what would you say if we added smth like this:


```typescript
async function runOrLog<T>(cb: () => Promise<T>, defaultValue: T): Promise<T> {
// or async function runOrLog<T>(cb: () => Promise<T>, defaultValue: T): Promise<T> {
  try {
    return await cb();
  } catch (error) {
    console.error("Error occurred:", error);
    // or report to sentry

    return defaultValue;
  }
}

// to later be used like
const { now, compare } = await runOrLog(getDashboardData, {})
const networkCapacityStats = await runOrLog(getProviderGraphData("count"), 0);
```

it would clean the code up quite a bit

File: apps/api/src/routes/v1/dashboardData.ts
Comment (ygrishajev) [inline apps/api/src/routes/v1/dashboardData.ts]: if you stick to this kind of declarations, please type them to avoid possible issues in the future

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: the purpose of default declarations is kinda obvious
```suggestion
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: these vars should be typed according to what would be assigned further

File: apps/stats-web/src/app/(home)/DashboardContainer.tsx
Comment (ygrishajev) [inline apps/stats-web/src/app/(home)/DashboardContainer.tsx]:
```suggestion
              Last updated: {!!dashboardData?.now?.date && <FormattedDate value={dashboardData.now.date} />} {!!dashboardData?.now?.date && <FormattedTime value={dashboardData.now.date} />}
```

File: apps/api/src/services/db/statsService.ts
Comment (Redm4x) [reply apps/api/src/services/db/statsService.ts]: I agree `prev`/`current` would be better names. This is part of the public api however, maybe we should hold off renaming those if it's not needed in case somebody is using it already? At least until we release a v2 of the api.

Comment (ygrishajev) [reply apps/api/src/services/db/statsService.ts]: Yea, agreed. This would be a breaking change, should be versioned. I missed these props were here before. Scratch my comment )

Comment (ygrishajev) [reply apps/api/src/services/db/statsService.ts:44]: I missed these props were here before. Scratch my comment )

Comment (forbesus) [reply apps/api/src/services/db/statsService.ts]: I agree with @Redm4x 
But if there is no one who are using `now`/`compare`, I will change

Comment (ygrishajev) [reply apps/api/src/services/db/statsService.ts]: We can't know for sure. This kind of changes should be versioned

**Comment by forbesus** [comment]
Hello @Redm4x  @ygrishajev 
I finished to fix
Would you check again?

File: apps/api/src/routes/v1/dashboardData.ts
Comment (ygrishajev) [inline apps/api/src/routes/v1/dashboardData.ts]:
```suggestion
const logger = new LoggerService({ context: "Dashboard" });
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: this call and below is really the same as in another file. Logger can be a parameter.

```typescript
const createLoggingExecutor = (logger: LoggerService) => async <T>(cb: () => Promise<T>, defaultValue?: T): Promise<T> => {
   try {
     return await cb();
   } catch (error) {
     logger.error({ event: `Failed to fetch ${cb.name}`, error });
     getSentry().captureException(error);
     return defaultValue;
   }
}

const logger = new LoggerService()
const runOrLog = createLoggingExecutor(logger)

const bondedTokens = await runOrLog(() => {
   const bondedTokensQuery = await axios.get<CosmosStakingPoolResponse>(
      `${apiNodeUrl}/cosmos/staking/v1beta1/pool`
   );
   return parseInt(bondedTokensQuery.data.pool.bonded_tokens);
})

// or
const bondedTokens = await runOrLog(
   () => axios.get<CosmosStakingPoolResponse>(`${apiNodeUrl}/cosmos/staking/v1beta1/pool`)
      .then(({ data }) => parseInt(data.pool.bonded_tokens))
)
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: this nested ternaries are really hard to understand. Could you please organize this into a flow with a couple of descriptive vars?

**Comment by forbesus** [comment]
Hello @ygrishajev I just fixed 
Would you check again?

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: what I meant is to extract this function into a dedicated module/file/service and then use it here and in another file

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]:
```suggestion
    stakingAPR,
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]:
```suggestion
  let stakingAPR: number | undefined;
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]:
```suggestion
  if (result.bondedTokens && result.bondedTokens > 0 && result.inflation && result.communityTax && result.totalSupply) {
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]:
```suggestion
```

**Comment by forbesus** [comment]
Hello @ygrishajev  How about now?

File: apps/api/src/routes/v1/dashboardData.ts
Comment (ygrishajev) [inline apps/api/src/routes/v1/dashboardData.ts]: what do you think about one last improvement here? Promise.all should speed up this endpoint quite a bit

```typescript
const [{ now, compare }, chainStatsQuery, networkCapacity, networkCapacityStats, latestBlocks, latestTransactions] = await Promise.all([
    runOrLog(getDashboardData),
    runOrLog(getChainStats, {
      bondedTokens: undefined,
      totalSupply: undefined,
      communityPool: undefined,
      inflation: undefined,
      stakingAPR: undefined
    }),
    runOrLog(getNetworkCapacity),
    runOrLog(() => getProviderGraphData("count")),
    runOrLog(() => getBlocks(5)),
    runOrLog(() => getTransactions(5))
  ]);

  const chainStats = {
    ...chainStatsQuery,
    height: latestBlocks && latestBlocks.length > 0 ? latestBlocks[0].height : undefined,
    transactionCount: latestBlocks && latestBlocks.length > 0 ? latestBlocks[0].totalTransactionCount : undefined
  };
```

File: apps/api/src/services/external/apiNodeService.ts
Comment (ygrishajev) [inline apps/api/src/services/external/apiNodeService.ts]: Same here, promise.all should be of good help

```typescript
const bondedTokensAsPromised = runOrLog(async () => {
        const bondedTokensQuery = await axios.get<CosmosStakingPoolResponse>(`${apiNodeUrl}/cosmos/staking/v1beta1/pool`);
        return parseInt(bondedTokensQuery.data.pool.bonded_tokens);
      });

      const totalSupplyAsPromised = runOrLog(async () => {
        const supplyQuery = await axios.get<CosmosBankSupplyResponse>(`${apiNodeUrl}/cosmos/bank/v1beta1/supply?pagination.limit=1000`);
        return parseInt(supplyQuery.data.supply.find(x => x.denom === "uakt")?.amount || "0");
      });

      const communityPoolAsPromised = runOrLog(async () => {
        const communityPoolQuery = await axios.get<CosmosDistributionCommunityPoolResponse>(`${apiNodeUrl}/cosmos/distribution/v1beta1/community_pool`);
        return parseFloat(communityPoolQuery.data.pool.find(x => x.denom === "uakt")?.amount || "0");
      });

      const inflationAsPromised = runOrLog(async () => {
        const inflationQuery = await axios.get<CosmosMintInflationResponse>(`${apiNodeUrl}/cosmos/mint/v1beta1/inflation`);
        return parseFloat(inflationQuery.data.inflation || "0");
      });

      const communityTaxAsPromised = runOrLog(async () => {
        const distributionQuery = await axios.get<CosmosDistributionParamsResponse>(`${apiNodeUrl}/cosmos/distribution/v1beta1/params`);
        return parseFloat(distributionQuery.data.params.community_tax || "0");
      });

      const [bondedTokens, totalSupply, communityPool, inflation, communityTax] = await Promise.all([
        bondedTokensAsPromised,
        totalSupplyAsPromised,
        communityPoolAsPromised,
        inflationAsPromised,
        communityTaxAsPromised
      ]);

      return {
        communityPool,
        inflation,
        communityTax,
        bondedTokens,
        totalSupply
      };
```

**Review by ygrishajev — APPROVED** [review]
Thanks for all your changes @forbesus! This looks so much better!

I'm adding a couple of comments, that you may improve further if you like. Anyways lgtm.
Let's please make sure it's also good with @Redm4x

**Comment by ygrishajev** [comment]
oh, 1 thing before we can merge. Please squash the commits and use a scope for the resulting commits. e.g.

```
feat(stats): ...message
```

This should fail on pre-commit hook coz this scope is not added to our [rules](https://github.com/akash-network/console/blob/73ff83843947c16a75085b7d7e2ddcead49bc67a/.commitlintrc.json). Please feel free to add it.

@Redm4x does this scope make sense to you?

**Comment by forbesus** [comment]
All things are finished @Redm4x  cc: @ygrishajev

**Comment by ygrishajev** [comment]
We follow rebase strategy, can you squash the merge commit too pls?

**Comment by forbesus** [comment]
How about now? @ygrishajev  cc: @Redm4x

**Review by ygrishajev — APPROVED** [review]
Lgtm! Thanks for all your changes!

**Review by Redm4x — APPROVED** [review]

**Pull request merged**
